FocalBlur

danieldresser-ie opened the pull request:
Putting this up for initial review. I'm guessing that the main thing we might iterate on is the interface.

I think I've got it correctly matching rendered DOF based on any reasonable camera parameters, though it would be worth double checking this if you've got any good examples of weird cameras. ( The only camera parameter I'm aware of that I'm not matching is if the aperture doesn't match the aspect ratio of the image resolution, and the film fit is not horizontal ... most people just use a horizontal fit, I think it might be worth just using horizontal aperture for simplicity? If someone is using a vertical fit with a weird image size they can get a match just by using the blurMultiplier to dial in the image scaling ).

My currently known outstanding work: renaming nodes, adding some documentation to backdrops, add handling to the infilling algorithm for the edge of the image ( currently, the edge of the image produces some weird effects ).

Perhaps the biggest uncertainty with this as is: what should the default value for `infilling` be? In my testing, it's been consistently improving quality, but is currently rather slow, and the approach is quite heuristic driven rather than being canonical ... I have some thoughts, we can talk about it tomorrow.

murraystevenson commented on python/GafferOSL/FocalBlur.gfr:
```diff
@@ -0,0 +1,889 @@
+__children["FocalBlur"]["Dot25"]["__uiPosition"].setValue( imath.V2f( 822.9505, 85.5897293 ) )
+__children["FocalBlur"]["Dot26"]["in"].setInput( __children["FocalBlur"]["Dot25"]["out"] )
+__children["FocalBlur"]["Dot26"]["__uiPosition"].setValue( imath.V2f( 822.9505, 35.2791519 ) )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["cameraPath"], 'nodule:type', '' )
```
Maybe we should use a `GafferSceneUI.ScenePathPlugValueWidget` here so users can browse for the camera? I think this may throw a small spanner as far as the placeholderText goes...

danieldresser-ie replied:
I've made this change, though, yes, it does throw a spanner in the works as far as placeholderText is concerned.

murraystevenson commented on python/GafferOSL/FocalBlur.gfr:
```diff
@@ -0,0 +1,889 @@
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["horizontalAperture"], 'layout:index', 16 )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["horizontalAperture"], 'nodule:type', '' )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["horizontalAperture"], 'layout:visibilityActivator', 'cameraCustom' )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["horizontalAperture"], 'description', "The aperture is one of the primary attributes of a real lens ( driven by cameraScene ).\n\nIn conjunction with the focal length, this defines the camera's field of view. For the purpose of computing focal blur, we just take the horizontal aperture, which we expect to correspond to the horizontal resoluti ...
```
This description includes "( driven by cameraScene )" while the one for `computedHorizontalAperture` does not...

murraystevenson replied:
Think it'd be worth describing that "aperture" here refers to the horizontal film-back or sensor width rather than a primary attribute of a real lens (where it may be confused with f-stop?).

murraystevenson commented on python/GafferOSL/FocalBlur.gfr:
```diff
@@ -0,0 +1,889 @@
+__children["FocalBlur"]["Backdrop15"]["title"].setValue( 'Apply Infill If Enabled' )
+__children["FocalBlur"]["Backdrop15"]["__uiPosition"].setValue( imath.V2f( 825.123169, -318.060547 ) )
+__children["FocalBlur"]["Backdrop15"]["__uiBound"].setValue( imath.Box2f( imath.V2f( -8.5, -5.08203125 ), imath.V2f( 45.1524048, 11.0820312 ) ) )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["cameraMode"], 'nodule:type', '' )
```
`cameraMode` requires a description.

murraystevenson commented on python/GafferOSL/FocalBlur.gfr:
```diff
@@ -0,0 +1,889 @@
+__children["FocalBlur"]["Backdrop14"]["title"].setValue( 'Delete Temporary Channels' )
+__children["FocalBlur"]["Backdrop14"]["__uiPosition"].setValue( imath.V2f( 825.123169, -345.381042 ) )
+__children["FocalBlur"]["Backdrop14"]["__uiBound"].setValue( imath.Box2f( imath.V2f( -9.19067383, -5.08203125 ), imath.V2f( 30.4585571, 11.0820312 ) ) )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["infilling"], 'description', "Enable additional blur to infill gaps.\n\nWhen there is a sharp variation in the amount of blur between different pixels, gaps are created because the algorithm doesn't know what was behind something that has been blurred out. For example, consider a small object that is completely out of focus in front of an in- ...
```
"using a series of blurs that try estimate what colors"...

murraystevenson commented on python/GafferOSL/FocalBlur.gfr:
```diff
@@ -0,0 +1,889 @@
+__children["FocalBlur"].addChild( Gaffer.Dot( "Dot16" ) )
+__children["FocalBlur"]["Dot16"].setup( GafferImage.ImagePlug( "in", ) )
+__children["FocalBlur"]["Dot16"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
+__children["FocalBlur"].addChild( Gaffer.FloatPlug( "backgroundZValue", defaultValue = 100000.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
```
Seeing as we've renamed the other "Z" plugs to "Depth", should this be "backgroundDepthValue" or "backgroundValue"?

murraystevenson commented on python/GafferOSL/FocalBlur.gfr:
```diff
@@ -0,0 +1,889 @@
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["backgroundZValue"], 'layout:index', 20 )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["backgroundZValue"], 'nodule:type', '' )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["backgroundZValue"], 'layout:visibilityActivator', 'depthInterpretationFilteredWithBackground' )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["backgroundZValue"], 'description', 'Set this if your renderer writes a constant Z value in empty pixels that we need to subtract off before we can upremultiply the edges. In Arnold this comes from the far clip plane of the camera you rendered with.' )
```
Maybe this description would be more clear along these lines?

The value to be subtracted before the depth channel is unpremultiplied. When rendering filtered depth from Arnold, this value is the far clip plane of the render camera.

danieldresser-ie replied:
Hmm, I still think that mentioning how you would identify an image where this is needed is the first thing a user needs to know about this value. I've shifted to using shorter sentences more in line with your wording, though:

`Set this if your renderer writes a constant Z value in empty pixels. This value will be subtracted before the depth channel is unpremultiplied. When rendering filtered depth from Arnold, this value is the far clip plane of the render camera.`

danieldresser-ie replied:
Oh, I see you've moved that part into the description for depthInterpretation, sure, that makes sense, I'll just go with your wording.

murraystevenson commented on python/GafferOSL/FocalBlur.gfr:
```diff
@@ -0,0 +1,889 @@
+__children["FocalBlur"]["Dot14"]["__uiPosition"].setValue( imath.V2f( 940.699768, 5.03942108 ) )
+__children["FocalBlur"]["focusDistance"].setValue( 70.0 )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["focusDistance"], 'noduleLayout:section', 'left' )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["focusDistance"], 'description', 'The Z value that is perfectly in focus.' )
```
We should get rid of the references to "Z" here and in `computedFocusDistance`'s description.

murraystevenson commented on python/GafferOSL/FocalBlur.gfr:
```diff
@@ -0,0 +1,889 @@
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["depthInterpretation"], 'preset:Premultiplied', 'premultiplied' )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["depthInterpretation"], 'preset:Filtered With Background Value ( Arnold )', 'filteredWithBackground' )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["depthInterpretation"], 'layout:index', 19 )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["depthInterpretation"], 'description', 'When using the depth channel, we need to know how it was stored? Is it premultiplied by the alpha ( typical when is set to write a depth image with a regular pixel filter )? Not premultiplied ( typical when using a pixel filter like "closest" )? Blended together with a weird massive background value set ...
```
I don't mind `depthInterpretation` as a plug name, though maybe the description could be improved by approaching it roughly along these lines?

```
How the depth channel should be interpreted :

- Premultiplied : Depth has been premultiplied by the alpha (typical when writing depth with a regular pixel filter).
- Unpremultiplied : Depth has not been premultiplied by the alpha (typical when writing depth with a pixel filter such as "closest").
- Premultiplied with background value (Arnold) : Depth includes a constant value for empty pixels, which needs to be subtracted before unpremultiplying. (typical with Arnold filtered depth which includes the camera far clip plane).
```

murraystevenson commented on python/GafferOSL/FocalBlur.gfr:
```diff
@@ -0,0 +1,889 @@
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["in"], 'noduleLayout:spacing', 2.0 )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["in"], 'layout:section', 'Settings' )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["in"], 'layout:index', 0 )
+__children["FocalBlur"]["out"].setInput( __children["FocalBlur"]["BoxOut"]["__out"] )
```
We're somehow serialising these `setInput()` calls, I noticed it when loading a script saved with the previous iteration of FocalBlur and saw errors due to the removal of the BoxOuts in this update.

danieldresser-ie replied:
This seems to be a general problem with ExtensionAlgo based nodes. I see the same problem with MetadataOverlay.

murraystevenson replied:
Thanks, after discussing offline we're happy to punt on this for now. This will need to be more generally solved at a later date.

murraystevenson reviewed (commented):
Thanks Daniel! I haven't gone too far into using this iteration of FocalBlur yet, but to begin with here are a few thoughts on the UI...

danieldresser-ie commented:
I've addressed Murray's feedback, and done a bunch of clean up and documentation.

I've also added a little bit of correction to reduce the weirdness at data window boundaries. This doesn't fully correct the issue, but it's a definite improvement. I'm worried that if I turn up the hackery too much, it will start causing other problems. I don't think it will be possible to achieve a truly correct solution to this problem without a variable radius "gathering" blur, but the current bag of heuristics seem to consistently be giving somewhat reasonable results.

I've also added a Changelog entry - as far as I know, this should be good to merge if you're happy with it.

murraystevenson commented on python/GafferOSL/FocalBlur.gfr:
```diff
@@ -0,0 +1,936 @@
+parent["variables"].addChild( Gaffer.NameValuePlug( "project:name", Gaffer.StringPlug( "value", defaultValue = 'default', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), "projectName", Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
+parent["variables"].addChild( Gaffer.NameValuePlug( "project:rootDirectory", Gaffer.StringPlug( "value", defaultValue = '$HOME/gaffer/projects/${project:name}', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), "projectRootDirectory", Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
+parent["variables"].addChild( Gaffer.NameValuePlug( "renderPass", Gaffer.StringPlug( "value", defaultValue = '', ), "renderPass", Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
+__children["FocalBlur"] = Gaffer.Box( "FocalBlur" )
```
The FocalBlur node itself will also require a description otherwise it returns "Holds a nested node graph of its own." from SubGraph.

murraystevenson commented on python/GafferOSL/FocalBlur.gfr:
```diff
@@ -0,0 +1,936 @@
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["in"], 'layout:section', 'Settings' )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["in"], 'layout:index', 0 )
+__children["FocalBlur"]["out"].setInput( __children["FocalBlur"]["BoxOut"]["__out"] )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["out"], 'description', 'The final result of the loop.' )
```
"The final result of the loop." -> "The output image generated by this node."

murraystevenson commented on python/GafferOSL/FocalBlur.gfr:
```diff
@@ -0,0 +1,936 @@
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["depthChannel"], 'nodule:type', '' )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["depthChannel"], 'layout:section', 'Depth' )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["depthChannel"], 'layout:index', 18 )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["depthChannel"], 'description', 'Which image channel to use as the depth when computing the amount of blur.' )
```
Would be user-friendly to make use of a `GafferImageUI.ChannelPlugValueWidget` for `depthChannel`.

murraystevenson commented on python/GafferOSL/FocalBlur.gfr:
```diff
@@ -0,0 +1,936 @@
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["depthInterpretation"], 'preset:Premultiplied', 'premultiplied' )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["depthInterpretation"], 'preset:Filtered With Background Value ( Arnold )', 'filteredWithBackground' )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["depthInterpretation"], 'layout:index', 19 )
+Gaffer.Metadata.registerValue( __children["FocalBlur"]["depthInterpretation"], 'description', 'How the depth channel should be interpreted :\n\n- Premultiplied : Depth has been premultiplied by the alpha (typical when writing depth with a regular pixel filter).\n- Unpremultiplied : Depth has not been premultiplied by the alpha (typical when writing depth with a pixel filter such as "closest").\n- ...
```
We've ended up with "Premultiplied with background value (Arnold)" in this description, and "Filtered With Background Value ( Arnold )" in the preset name. We should standardise on one or the other.

murraystevenson reviewed (commented):
Thanks for the updates, I've taken another look at the UI and the resulting Node Reference entry in the documentation and spotted a few more issues. Feel free to squash them down so we're ready to merge.
